File: Health.lua
--[[--------------------------------------------------------------------
    Copyright (C) 2015 Johnny C. Lam.
    See the file LICENSE.txt for copying permission.
--]]--------------------------------------------------------------------

local OVALE, Ovale = ...
local OvaleHealth = Ovale:NewModule("OvaleHealth", "AceEvent-3.0")
Ovale.OvaleHealth = OvaleHealth

--<private-static-properties>
local OvaleDebug = Ovale.OvaleDebug
local OvaleProfiler = Ovale.OvaleProfiler

-- Forward declarations for module dependencies.
local OvaleData = nil
local OvaleGUID = nil
local OvaleState = nil

local strsub = string.sub
local tonumber = tonumber
local wipe = wipe
local API_GetTime = GetTime
local API_UnitHealth = UnitHealth
local API_UnitHealthMax = UnitHealthMax
local INFINITY = math.huge

-- Register for debugging messages.
OvaleDebug:RegisterDebugging(OvaleHealth)
-- Register for profiling.
OvaleProfiler:RegisterProfiling(OvaleHealth)

local CLEU_DAMAGE_EVENT = {
	DAMAGE_SHIELD = true,
	DAMAGE_SPLIT = true,
	RANGE_DAMAGE = true,
	SPELL_BUILDING_DAMAGE = true,
	SPELL_DAMAGE = true,
	SPELL_PERIODIC_DAMAGE = true,
	SWING_DAMAGE = true,
	ENVIRONMENTAL_DAMAGE = true,
}

local CLEU_HEAL_EVENT = {
	SPELL_HEAL = true,
	SPELL_PERIODIC_HEAL = true,
}
--</private-static-properties>

--<public-static-properties>
-- Health of unit, indexed by GUID.
OvaleHealth.health = {}
-- Maximum health of unit, indexed by GUID.
OvaleHealth.maxHealth = {}
-- Running total of damage taken, indexed by GUID.
OvaleHealth.totalDamage = {}
-- Running total of healing taken, indexed by GUID.
OvaleHealth.totalHealing = {}
-- The CLEU timestamp that a GUID first was seen taking damage or healing.
OvaleHealth.firstSeen = {}
-- The CLEU timestamp that a GUID most recently was seen taking damage or healing.
OvaleHealth.lastUpdated = {}
-- Unused public property to suppress lint warnings.
--OvaleHealth.defaultTarget = nil
--</public-static-properties>

--<public-static-methods>
function OvaleHealth:OnInitialize()
	-- Resolve module dependencies.
	OvaleData = Ovale.OvaleData
	OvaleGUID = Ovale.OvaleGUID
	OvaleState = Ovale.OvaleState
end

function OvaleHealth:OnEnable()
	self:RegisterEvent("PLAYER_REGEN_DISABLED")
	self:RegisterEvent("PLAYER_REGEN_ENABLED")
	self:RegisterEvent("UNIT_HEALTH_FREQUENT", "UpdateHealth")
	self:RegisterEvent("UNIT_MAXHEALTH", "UpdateHealth")
	self:RegisterMessage("Ovale_UnitChanged")
	OvaleData:RegisterRequirement("health_pct", "RequireHealthPercentHandler", self)
	OvaleData:RegisterRequirement("pet_health_pct", "RequireHealthPercentHandler", self)
	OvaleData:RegisterRequirement("target_health_pct", "RequireHealthPercentHandler", self)
	OvaleState:RegisterState(self, self.statePrototype)
end

function OvaleHealth:OnDisable()
	OvaleState:UnregisterState(self)
	OvaleData:UnregisterRequirement("health_pct")
	OvaleData:UnregisterRequirement("pet_health_pct")
	OvaleData:UnregisterRequirement("target_health_pct")
	self:UnregisterEvent("PLAYER_REGEN_ENABLED")
	self:UnregisterEvent("PLAYER_TARGET_CHANGED")
	self:UnregisterEvent("UNIT_HEALTH_FREQUENT")
	self:UnregisterEvent("UNIT_MAXHEALTH")
	self:UnregisterMessage("Ovale_UnitChanged")
end

function OvaleHealth:COMBAT_LOG_EVENT_UNFILTERED(event, timestamp, cleuEvent, hideCaster, sourceGUID, sourceName, sourceFlags, sourceRaidFlags, destGUID, destName, destFlags, destRaidFlags, ...)
	local arg12, arg13, arg14, arg15, arg16, arg17, arg18, arg19, arg20, arg21, arg22, arg23, arg24, arg25 = ...
	self:StartProfiling("OvaleHealth_COMBAT_LOG_EVENT_UNFILTERED")
	-- Keep a running total of damage and healing taken per GUID.
	local healthUpdate = false
	if CLEU_DAMAGE_EVENT[cleuEvent] then
		local amount
		if cleuEvent == "SWING_DAMAGE" then
			amount = arg12
		elseif cleuEvent == "ENVIRONMENTAL_DAMAGE" then
			amount = arg13
		else
			amount = arg15
		end
		self:Debug(cleuEvent, destGUID, amount)
		local total = self.totalDamage[destGUID] or 0
		self.totalDamage[destGUID] = total + amount
		healthUpdate = true
	elseif CLEU_HEAL_EVENT[cleuEvent] then
		local amount = arg15
		self:Debug(cleuEvent, destGUID, amount)
		local total = self.totalHealing[destGUID] or 0
		self.totalHealing[destGUID] = total + amount
		healthUpdate = true
	end
	if healthUpdate then
		if not self.firstSeen[destGUID] then
			self.firstSeen[destGUID] = timestamp
		end
		self.lastUpdated[destGUID] = timestamp
	end
	self:StopProfiling("OvaleHealth_COMBAT_LOG_EVENT_UNFILTERED")
end

function OvaleHealth:PLAYER_REGEN_DISABLED(event)
	-- Entering combat.
	self:RegisterEvent("COMBAT_LOG_EVENT_UNFILTERED")
end

function OvaleHealth:PLAYER_REGEN_ENABLED(event)
	-- Leaving combat.
	self:UnregisterEvent("COMBAT_LOG_EVENT_UNFILTERED")
	-- Clear running totals when leaving combat.
	wipe(self.totalDamage)
	wipe(self.totalHealing)
	wipe(self.firstSeen)
	wipe(self.lastUpdated)
end

function OvaleHealth:Ovale_UnitChanged(event, unitId, guid)
	self:StartProfiling("Ovale_UnitChanged")
	if unitId == "target" or unitId == "focus" then
		self:Debug(event, unitId, guid)
		self:UpdateHealth("UNIT_HEALTH_FREQUENT", unitId)
		self:UpdateHealth("UNIT_MAXHEALTH", unitId)
		self:StopProfiling("Ovale_UnitChanged")
	end
end

function OvaleHealth:UpdateHealth(event, unitId)
	if not unitId then return end
	self:StartProfiling("OvaleHealth_UpdateHealth")
	local func = API_UnitHealth
	local db = self.health
	if event == "UNIT_MAXHEALTH" then
		func = API_UnitHealthMax
		db = self.maxHealth
	end
	local amount = func(unitId)
	if amount then
		local guid = OvaleGUID:UnitGUID(unitId)
		self:Debug(event, unitId, guid, amount)
		if guid then
			if amount > 0 then
				db[guid] = amount
			else
				db[guid] = nil
				self.firstSeen[guid] = nil
				self.lastUpdated[guid] = nil
			end
			Ovale.refreshNeeded[guid] = true
		end
	end
	self:StopProfiling("OvaleHealth_UpdateHealth")
end

--[[
	Return the current health of the unit.
	An optional GUID may be passed as a hint for the GUID of the unit.
--]]
function OvaleHealth:UnitHealth(unitId, guid)
	local amount
	if unitId then
		guid = guid or OvaleGUID:UnitGUID(unitId)
		if guid then
			if unitId == "target" or unitId == "focus" then
				-- The target and focus target are actively tracked.
				amount = self.health[guid] or 0
			else
				-- Cache the health for later reference.
				amount = API_UnitHealth(unitId)
				self.health[guid] = amount
			end
		else
			amount = 0
		end
	end
	return amount
end

--[[
	Return the maximum health of the unit.
	An optional GUID may be passed as a hint for the GUID of the unit.
--]]
function OvaleHealth:UnitHealthMax(unitId, guid)
	local amount
	if unitId then
		guid = guid or OvaleGUID:UnitGUID(unitId)
		if guid then
			if unitId == "target" or unitId == "focus" then
				-- The target and focus target are actively tracked.
				amount = self.maxHealth[guid] or 0
			else
				-- Cache the maximum health for later reference.
				amount = API_UnitHealthMax(unitId)
				self.maxHealth[guid] = amount
			end
		else
			amount = 0
		end
	end
	return amount
end

--[[
	Return the estimated time to die in seconds for the unit.
	An optional GUID may be passed as a hint for the GUID of the unit.
--]]
function OvaleHealth:UnitTimeToDie(unitId, guid)
	self:StartProfiling("OvaleHealth_UnitTimeToDie")
	local timeToDie = INFINITY
	guid = guid or OvaleGUID:UnitGUID(unitId)
	if guid then
		local health = self:UnitHealth(unitId, guid)
		local maxHealth = self:UnitHealthMax(unitId, guid)
		if health and maxHealth then
			if health == 0 then
				timeToDie = 0
				self.firstSeen[guid] = nil
				self.lastUpdated[guid] = nil
			elseif maxHealth > 5 then
				-- Filter out targets whose maximum health is less than 5, which are probably target dummies.
				local firstSeen, lastUpdated = self.firstSeen[guid], self.lastUpdated[guid]
				local damage = self.totalDamage[guid] or 0
				local healing = self.totalHealing[guid] or 0
				if firstSeen and lastUpdated and lastUpdated > firstSeen and damage > healing then
					timeToDie = health * (lastUpdated - firstSeen) / (damage - healing)
				end
			end
		end
	end
	self:StopProfiling("OvaleHealth_UnitTimeToDie")
	return timeToDie
end

-- Run-time check that the target is below a health percent threshold.
-- NOTE: Mirrored in statePrototype below.
function OvaleHealth:RequireHealthPercentHandler(spellId, atTime, requirement, tokens, index, targetGUID)
	local verified = false
	-- If index isn't given, then tokens holds the actual token value.
	local threshold = tokens
	if index then
		threshold = tokens[index]
		index = index + 1
	end
	if threshold then
		local isBang = false
		if strsub(threshold, 1, 1) == "!" then
			isBang = true
			threshold = strsub(threshold, 2)
		end
		threshold = tonumber(threshold) or 0
		local guid, unitId
		if strsub(requirement, 1, 7) == "target_" then
			if targetGUID then
				guid = targetGUID
				unitId = OvaleGUID:GUIDUnit(guid)
			else
				unitId = self.defaultTarget or "target"
			end
		elseif strsub(requirement, 1, 4) == "pet_" then
			unitId = "pet"
		else
			unitId = "player"
		end
		guid = guid or OvaleGUID:UnitGUID(unitId)
		local health = OvaleHealth:UnitHealth(unitId, guid) or 0
		local maxHealth = OvaleHealth:UnitHealthMax(unitId, guid) or 0
		local healthPercent = (maxHealth > 0) and (health / maxHealth * 100) or 100
		if not isBang and healthPercent <= threshold or isBang and healthPercent > threshold then
			verified = true
		end
		local result = verified and "passed" or "FAILED"
		if isBang then
			self:Log("    Require %s health > %f%% (%f) at time=%f: %s", unitId, threshold, healthPercent, atTime, result)
		else
			self:Log("    Require %s health <= %f%% (%f) at time=%f: %s", unitId, threshold, healthPercent, atTime, result)
		end
	else
		Ovale:OneTimeMessage("Warning: requirement '%s' is missing a threshold argument.", requirement)
	end
	return verified, requirement, index
end
--</public-static-methods>

--[[----------------------------------------------------------------------------
	State machine for simulator.
--]]----------------------------------------------------------------------------

--<public-static-properties>
OvaleHealth.statePrototype = {}
--</public-static-properties>

--<private-static-properties>
local statePrototype = OvaleHealth.statePrototype
--</private-static-properties>

--<state-methods>
-- Mirrored methods.
statePrototype.RequireHealthPercentHandler = OvaleHealth.RequireHealthPercentHandler
--</state-methods>

File: src/Health.ts
import { OvaleGUIDClass } from "./GUID";
import { Tokens, OvaleRequirement } from "./Requirement";
import aceEvent, { AceEvent } from "@wowts/ace_event-3.0";
import { sub } from "@wowts/string";
import { tonumber, wipe, LuaObj } from "@wowts/lua";
import { UnitHealth, UnitHealthMax, UnitGetTotalAbsorbs, UnitGetTotalHealAbsorbs, CombatLogGetCurrentEventInfo } from "@wowts/wow-mock";
import { huge } from "@wowts/math";
import { BaseState } from "./BaseState";
import { OvaleClass } from "./Ovale";
import { AceModule } from "@wowts/tsaddon";
import { OvaleOptionsClass } from "./Options";
import { Profiler, OvaleProfilerClass } from "./Profiler";
import { Tracer, OvaleDebugClass } from "./Debug";


let INFINITY = huge;
let CLEU_DAMAGE_EVENT = {
    DAMAGE_SHIELD: true,
    DAMAGE_SPLIT: true,
    RANGE_DAMAGE: true,
    SPELL_BUILDING_DAMAGE: true,
    SPELL_DAMAGE: true,
    SPELL_PERIODIC_DAMAGE: true,
    SWING_DAMAGE: true,
    ENVIRONMENTAL_DAMAGE: true
}
let CLEU_HEAL_EVENT = {
    SPELL_HEAL: true,
    SPELL_PERIODIC_HEAL: true
}

export class OvaleHealthClass {
    health: LuaObj<number> = {}
    maxHealth: LuaObj<number> = {}
    absorb: LuaObj<number> = {}
    healAbsorb: LuaObj<number> = {}
    totalDamage: LuaObj<number> = {}
    totalHealing: LuaObj<number> = {}
    firstSeen: LuaObj<number> = {}
    lastUpdated: LuaObj<number> = {}
    private module: AceModule & AceEvent;
    private profiler: Profiler;
    private tracer: Tracer;

    constructor(
        private ovaleGuid: OvaleGUIDClass,
        private baseState: BaseState,
        private ovale: OvaleClass,
        private ovaleOptions: OvaleOptionsClass,
        ovaleDebug: OvaleDebugClass,
        ovaleProfiler: OvaleProfilerClass,
        private requirement: OvaleRequirement
    ) {
        
        this.module = ovale.createModule("OvaleHealth", this.OnInitialize, this.OnDisable, aceEvent);
        this.tracer = ovaleDebug.create(this.module.GetName());
        this.profiler = ovaleProfiler.create(this.module.GetName());
    }

    private OnInitialize = () => {
        this.module.RegisterEvent("PLAYER_REGEN_DISABLED", this.PLAYER_REGEN_DISABLED);
        this.module.RegisterEvent("PLAYER_REGEN_ENABLED", this.PLAYER_REGEN_ENABLED);
        if (this.ovaleOptions.db.profile.apparence.frequentHealthUpdates) {
            this.module.RegisterEvent("UNIT_HEALTH_FREQUENT", this.UpdateHealth);
        } else {
            this.module.RegisterEvent("UNIT_HEALTH", this.UpdateHealth);
        }
        this.module.RegisterEvent("UNIT_MAXHEALTH", this.UpdateHealth);
        this.module.RegisterEvent("UNIT_ABSORB_AMOUNT_CHANGED", this.UpdateAbsorb);
        this.module.RegisterEvent("UNIT_HEAL_ABSORB_AMOUNT_CHANGED", this.UpdateAbsorb);
        this.module.RegisterMessage("Ovale_UnitChanged", this.Ovale_UnitChanged);
        this.requirement.RegisterRequirement("health_pct", this.RequireHealthPercentHandler);
        this.requirement.RegisterRequirement("pet_health_pct", this.RequireHealthPercentHandler);
        this.requirement.RegisterRequirement("target_health_pct", this.RequireHealthPercentHandler);
    }
    private OnDisable = () => {
        this.requirement.UnregisterRequirement("health_pct");
        this.requirement.UnregisterRequirement("pet_health_pct");
        this.requirement.UnregisterRequirement("target_health_pct");
        this.module.UnregisterEvent("PLAYER_REGEN_ENABLED");
        this.module.UnregisterEvent("PLAYER_TARGET_CHANGED");
        this.module.UnregisterEvent("UNIT_HEALTH_FREQUENT");
        this.module.UnregisterEvent("UNIT_HEALTH");
        this.module.UnregisterEvent("UNIT_MAXHEALTH");
        this.module.UnregisterEvent("UNIT_ABSORB_AMOUNT_CHANGED");
        this.module.UnregisterEvent("UNIT_HEAL_ABSORB_AMOUNT_CHANGED");
        this.module.UnregisterMessage("Ovale_UnitChanged");
    }
    private COMBAT_LOG_EVENT_UNFILTERED = (event: string, ...__args: any[]) => {
        let [timestamp, cleuEvent, , , , , , destGUID, , , , arg12, arg13, , arg15] = CombatLogGetCurrentEventInfo();
        this.profiler.StartProfiling("OvaleHealth_COMBAT_LOG_EVENT_UNFILTERED");
        let healthUpdate = false;
        if (CLEU_DAMAGE_EVENT[<keyof typeof CLEU_DAMAGE_EVENT> cleuEvent]) {
            let amount;
            if (cleuEvent == "SWING_DAMAGE") {
                amount = arg12;
            } else if (cleuEvent == "ENVIRONMENTAL_DAMAGE") {
                amount = arg13;
            } else {
                amount = arg15;
            }
            this.tracer.Debug(cleuEvent, destGUID, amount);
            let total = this.totalDamage[destGUID] || 0;
            this.totalDamage[destGUID] = total + amount;
            healthUpdate = true;
        } else if (CLEU_HEAL_EVENT[<keyof typeof CLEU_HEAL_EVENT>cleuEvent]) {
            let amount = arg15;
            this.tracer.Debug(cleuEvent, destGUID, amount);
            let total = this.totalHealing[destGUID] || 0;
            this.totalHealing[destGUID] = total + amount;
            healthUpdate = true;
        }
        if (healthUpdate) {
            if (!this.firstSeen[destGUID]) {
                this.firstSeen[destGUID] = timestamp;
            }
            this.lastUpdated[destGUID] = timestamp;
        }
        this.profiler.StopProfiling("OvaleHealth_COMBAT_LOG_EVENT_UNFILTERED");
    }
    private PLAYER_REGEN_DISABLED = (event: string) => {
        this.module.RegisterEvent("COMBAT_LOG_EVENT_UNFILTERED", this.COMBAT_LOG_EVENT_UNFILTERED);
    }
    private PLAYER_REGEN_ENABLED = (event: string) => {
        this.module.UnregisterEvent("COMBAT_LOG_EVENT_UNFILTERED");
        wipe(this.totalDamage);
        wipe(this.totalHealing);
        wipe(this.firstSeen);
        wipe(this.lastUpdated);
    }
    private Ovale_UnitChanged = (event: string, unitId: string, guid: string) => {
        this.profiler.StartProfiling("Ovale_UnitChanged");
        if (unitId == "target" || unitId == "focus") {
            this.tracer.Debug(event, unitId, guid);
            this.UpdateHealth("UNIT_HEALTH_FREQUENT", unitId);
            this.UpdateHealth("UNIT_MAXHEALTH", unitId);
            this.UpdateAbsorb("UNIT_ABSORB_AMOUNT_CHANGED", unitId);
            this.UpdateAbsorb("UNIT_HEAL_ABSORB_AMOUNT_CHANGED", unitId);
        }
        this.profiler.StopProfiling("Ovale_UnitChanged");
    }
    private UpdateAbsorb = (event: string, unitId: string) => {
        if (!unitId) {
            return;
        }
        this.profiler.StartProfiling("OvaleHealth_UpdateAbsorb");
        
        let func;
        let db;
        
        if (event == "UNIT_ABSORB_AMOUNT_CHANGED") {
            func = UnitGetTotalAbsorbs;
            db = this.absorb;
        } else if (event == "UNIT_HEAL_ABSORB_AMOUNT_CHANGED") {
            func = UnitGetTotalHealAbsorbs;
            db = this.absorb;
        } else {
            this.ovale.OneTimeMessage("Warning: Invalid event (%s) in UpdateAbsorb.", event);
            return;
        }
        
        let amount = func(unitId);
        if(amount >= 0) {
            let guid = this.ovaleGuid.UnitGUID(unitId);
            this.tracer.Debug(event, unitId, guid, amount);
            if (guid) {
                db[guid] = amount;
            }
        }
        
        this.profiler.StopProfiling("OvaleHealth_UpdateHealth");
    }
    private UpdateHealth = (event: string, unitId: string) => {
        if (!unitId) {
            return;
        }
        this.profiler.StartProfiling("OvaleHealth_UpdateHealth");
        
        let func;
        let db;
        
        if (event == "UNIT_HEALTH_FREQUENT" || event == "UNIT_HEALTH") {
            func = UnitHealth;
            db = this.health;
        } else if (event == "UNIT_MAXHEALTH") {
            func = UnitHealthMax;
            db = this.maxHealth;
        } else {
            this.ovale.OneTimeMessage("Warning: Invalid event (%s) in UpdateHealth.", event);
            return;
        }
        
        let amount = func(unitId);
        if (amount) {
            let guid = this.ovaleGuid.UnitGUID(unitId);
            this.tracer.Debug(event, unitId, guid, amount);
            if (guid) {
                if (amount > 0) {
                    db[guid] = amount;
                } else {
                    db[guid] = undefined;
                    this.firstSeen[guid] = undefined;
                    this.lastUpdated[guid] = undefined;
                }
            }
        }
        this.profiler.StopProfiling("OvaleHealth_UpdateHealth");
    }
    UnitHealth(unitId: string, guid?: string) {
        return this.UnitAmount(UnitHealth, this.health, unitId, guid);
    }
    UnitHealthMax(unitId: string, guid?:string) {
        return this.UnitAmount(UnitHealthMax, this.maxHealth, unitId, guid);
    }
    UnitAbsorb(unitId: string, guid?: string) {
        return this.UnitAmount(UnitGetTotalAbsorbs, this.absorb, unitId, guid);
    }
    UnitHealAbsorb(unitId: string, guid?: string) {
        return this.UnitAmount(UnitGetTotalHealAbsorbs, this.healAbsorb, unitId, guid);
    }
    UnitAmount(func: (_: string) => number, db: LuaObj<number>, unitId: string, guid?: string): number {
        let amount;
        if (unitId) {
            guid = guid || this.ovaleGuid.UnitGUID(unitId);
            if (guid) {
                if (unitId == "target" || unitId == "focus") {
                    amount = db[guid] || 0;
                } else {
                    amount = func(unitId);
                    db[guid] = amount;
                }
            } else {
                amount = 0;
            }
        }
        return amount;
    }
    UnitTimeToDie(unitId: string, effectiveHealth?: boolean, guid?: string) {
        this.profiler.StartProfiling("OvaleHealth_UnitTimeToDie");
        let timeToDie = INFINITY;
        guid = guid || this.ovaleGuid.UnitGUID(unitId);
        if (guid) {
            let health = this.UnitHealth(unitId, guid);
            if (effectiveHealth) {
                health = health + this.UnitAbsorb(unitId, guid) - this.UnitHealAbsorb(unitId, guid);
            }
            let maxHealth = this.UnitHealthMax(unitId, guid);
            if (health && maxHealth) {
                if (health == 0) {
                    timeToDie = 0;
                    this.firstSeen[guid] = undefined;
                    this.lastUpdated[guid] = undefined;
                } else if (maxHealth > 5) {
                    let [firstSeen, lastUpdated] = [this.firstSeen[guid], this.lastUpdated[guid]];
                    let damage = this.totalDamage[guid] || 0;
                    let healing = this.totalHealing[guid] || 0;
                    if (firstSeen && lastUpdated && lastUpdated > firstSeen && damage > healing) {
                        timeToDie = health * (lastUpdated - firstSeen) / (damage - healing);
                    }
                }
            }
        }
        this.profiler.StopProfiling("OvaleHealth_UnitTimeToDie");
        return timeToDie;
    }
    RequireHealthPercentHandler = (spellId: number, atTime: number, requirement: string, tokens: Tokens, index: number, targetGUID: string): [boolean, string, number] => {
        let verified = false;
        let threshold = <string>tokens[index];
        index = index + 1;
        if (threshold) {
            let isBang = false;
            if (sub(threshold, 1, 1) == "!") {
                isBang = true;
                threshold = sub(threshold, 2);
            }
            const thresholdValue = tonumber(threshold) || 0;
            let guid, unitId: string;
            if (sub(requirement, 1, 7) == "target_") {
                if (targetGUID) {
                    guid = targetGUID;
                    [unitId] = this.ovaleGuid.GUIDUnit(guid);
                } else {
                    unitId = this.baseState.next.defaultTarget || "target";
                }
            } else if (sub(requirement, 1, 4) == "pet_") {
                unitId = "pet";
            } else {
                unitId = "player";
            }
            guid = guid || this.ovaleGuid.UnitGUID(unitId);
            let health = this.UnitHealth(unitId, guid) || 0;
            let maxHealth = this.UnitHealthMax(unitId, guid) || 1;
            let healthPercent = (health / maxHealth * 100) || 100;
            if (!isBang && healthPercent <= thresholdValue || isBang && healthPercent > thresholdValue) {
                verified = true;
            }
            let result = verified && "passed" || "FAILED";
            if (isBang) {
                this.tracer.Log("    Require %s health > %f%% (%f) at time=%f: %s", unitId, threshold, healthPercent, atTime, result);
            } else {
                this.tracer.Log("    Require %s health <= %f%% (%f) at time=%f: %s", unitId, threshold, healthPercent, atTime, result);
            }
        } else {
            this.ovale.OneTimeMessage("Warning: requirement '%s' is missing a threshold argument.", requirement);
        }
        return [verified, requirement, index];
    }
    CleanState(): void {
    }
    InitializeState(): void {
    }
    ResetState(): void {
    }
}

